### PR TITLE
feat: [MR-552] Define `RejectCode::SysUnknown` and error codes

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
[[MR-552]] Define new `RejectCode` variant `SysUnknown` (for when the response to a best-effort call is unknown) and two `ErrorCode` variants -- `DeadlineExpired` and `ResponseDropped` -- to be used when expiring best-effort calls or, respectively, when dropping best-effort responses.

[MR-552]: https://dfinity.atlassian.net/browse/MR-552?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by alin-at-dfinity** [comment]
Technically, we should also have a new certification version to go along with it (e.g. `V19` does not support `SysUnknown`, `V20` does), but the value is encoded as a plain `u8`, so it would basically only be "forward compatibility theater": i.e. it would only serve as a marker for "support for this needs to roll out to all subnets before it can be enabled on any subnet" (or the other way around).

I can do it, but it's definitely not necessary in any way. WDYT?

**Comment by rumenov** [comment]
@alin-at-dfinity one car argue that `DeadlineExpired` and `ResponseDropped` are SysTransient, right?

**Comment by stiegerc** [comment]
`RejectCode` is included in the reject response. The canister can access it using [this](https://sourcegraph.com/github.com/dfinity/ic@14e933e7077cbdfa6ae9b9926f8481b75d6c98fa/-/blob/rs/rust_canisters/dfn_core/src/api.rs?L713). I wonder whether this provides any useful information from the point of view of a canister, i.e. what does `SysUnknown` tell me that `SysTransient` doesn't in terms of how I should react?

I see you also defined new `ErrorCode` variants, which are use for the execution history. But we could also map those to `SysTransient` right?

**Comment by derlerd-dfinity** [comment]
@rumenov, @stiegerc we decided against using `SysTransient` because `Unknown` means it might not be safe to blindly retry (as the message might have reached the target canister) whereas `Transient` means it is safe to blindly retry [it is a separate, unrelated discussion whether or not blindly retrying is a good idea from an application perspective].

File: rs/types/types/src/exhaustive.rs [hunk 1]
```diff
@@ -284,7 +284,10 @@ impl<K: ExhaustiveSet + std::cmp::Ord> ExhaustiveSet for BTreeSet<K> {
 
 impl ExhaustiveSet for RejectCode {
     fn exhaustive_set<R: RngCore + CryptoRng>(_: &mut R) -> Vec<Self> {
-        RejectCode::iter().collect()
+        RejectCode::iter()
+            // TODO(MR-610): Drop this after `SysUnknown` is supported on mainnet.
+            .filter(|code| *code != RejectCode::SysUnknown)
```
Comment (derlerd-dfinity) [inline rs/types/types/src/exhaustive.rs:289]: I assume there is no safer way to ensure we don't forget to remove this than linking the ticket?

**Review by derlerd-dfinity — APPROVED** [review]
Code wise the change LGTM. 

I'd suggest to wait a bit to see whether there are strong objections against this new reject code based on the questions/discussions above.

**Review by oggy-dfin — APPROVED** [review]
LGTM, is in line with the proposed [interface spec change](https://github.com/dfinity/interface-spec/pull/268)

**Comment by oggy-dfin** [comment]
One thing I don't know is if we're still maintaining the [list of error codes](https://wiki.internetcomputer.org/wiki/Error_Codes_returned_by_Internet_Computer) - it's definitely gotten stale, but if we want to maintain it, we should add these codes there at some point.

**Comment by stiegerc** [comment]
> @rumenov, @stiegerc we decided against using `SysTransient` because `Unknown` means it might not be safe to blindly retry (as the message might have reached the target canister and was successfully executed) whereas `Transient` means it is safe to blindly retry [it is a separate, unrelated discussion whether or not blindly retrying is a good idea from an application perspective].

Ah I see, makes sense.

File: rs/types/types/src/exhaustive.rs (hunk 1 above)
Comment (stiegerc) [reply rs/types/types/src/exhaustive.rs:289]: I guess introducing a new certification version would help here. But I'm fine with this.

**Review by stiegerc — APPROVED** [review]

File: rs/types/types/src/exhaustive.rs (hunk 1 above)
Comment (alin-at-dfinity) [reply rs/types/types/src/exhaustive.rs:289]: That's what the test failure message suggested. I also don't think there's a safer way.

Comment (alin-at-dfinity) [reply rs/types/types/src/exhaustive.rs:289]: [Sorry, only replied to David above, GitHub did not render Christian's comment without refreshing.]

We can definitely add a new certification version, as said above. But it won't help with `ExhaustiveSet` and the CUP compatibility test. Or with remembering to drop this line after a couple of replica releases.

**Comment by rumenov** [comment]
> as the message might have reached the target canister and was successfully executed

Why if the message was successfully executed it cannot be retried ?

**Comment by stiegerc** [comment]
> > as the message might have reached the target canister and was successfully executed
> 
> Why if the message was successfully executed it cannot be retried ?

It can but it may still be important information for the canister that the message may have arrived, whereas with `SysTransient` you know for sure that it didn't. It stands to reason that you should use a guaranteed reply message for this kind of thing but that's a different discussion.

**Review by dsarlis — APPROVED** [review]

**Review by kpop-dfinity — APPROVED** [review]

**Pull request merged**